### PR TITLE
Support concentric neutral cables

### PR DESCRIPTION
**Pull request by veronicaguo**

### CODE REVIEW

File: tests/test_concentric_neutral_cable.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+def test_concentric_neutral_cable():
+    model = ConcentricNeutralCarsonsEquations(ABCCable())
+
+    assert_array_almost_equal(
+        model.impedance,
+        array([
+            [0.7981 + 1j*0.4467, 0.3188 + 1j*0.0334, 0.2848 + 1j*0.0138],
+            [0.3188 + 1j*0.0334, 0.7890 + 1j*0.4048, 0.3188 + 1j*0.0334],
+            [0.2848 + 1j*0.0138, 0.3188 + 1j*0.0334, 0.7981 + 1j*0.4467],
+
```
Comment (AnjoMan) [inline tests/test_concentric_neutral_cable.py]: extra space here

File: tests/test_concentric_neutral_cable.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+ureg = pint.UnitRegistry()
+
+feet = ureg.feet
+inches = ureg.inches
+miles = ureg.miles
+ohms = ureg.ohms
+
+
+def GMR_cn(GMR_s, k, R):
```
Comment (AnjoMan) [inline tests/test_concentric_neutral_cable.py]: i think we could drop this here

File: carsons/carsons.py [hunk 3]
```diff
@@ -183,3 +181,75 @@ def calculate_distance(positionᵢ, positionⱼ):
...
+        distance_ij = self.calculate_distance(self.phase_positions[i],
+                                              self.phase_positions[j])
+        if not I & J and 'N' in I ^ J:
+            # approximate by modelling the concentric neutral cables as one
+            # equivalent conductor directly above the phase conductor
+            return (distance_ij**2 + r**2) ** 0.5
+        else:
+            return distance_ij
+
+    def compute_P(self, i, j, number_of_terms=1):
```
Comment (etimberg) [inline carsons/carsons.py]: Should `number_of_terms` be passed through here?

File: carsons/carsons.py [hunk 4]
```diff
@@ -183,3 +181,75 @@ def calculate_distance(positionᵢ, positionⱼ):
     def get_h(self, i):
         _, yᵢ = self.phase_positions[i]
         return yᵢ
+
+    @property
+    def impedance(self):
```
Comment (etimberg) [inline carsons/carsons.py]: i think this new property should be tested. perhaps also deprecate the `convert_geometric_model` function

File: carsons/carsons.py [hunk 5]
```diff
@@ -183,3 +181,75 @@ def calculate_distance(positionᵢ, positionⱼ):
...
+        self.r.update({
+            phase: resistance / model.neutral_strand_count[phase]
+            for phase, resistance in model.neutral_strand_resistance.items()
+        })
+        return
+
+    def compute_d(self, i, j):
+        I, J = set(i), set(j)
+        r = self.radius[i] or self.radius[j]
+        if I ^ J == set('N') and 'N' not in I & J:
```
[the remark this answers is not in the document]
Comment (AnjoMan) [reply carsons/carsons.py]: ^ good idea. also think it might be worthwhile to assign the set evaluations to variables. for example `one_or_more_neutral = I^J == set('N')`, `both_neutral = 'N' in I & J`

File: carsons/carsons.py (hunk 4 above)
Comment (AnjoMan) [reply carsons/carsons.py]: it is called on the concentric neutral. its just added here because it doesn't depend on anything that `Carsons` doesn't have.

File: carsons/carsons.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (veronicaguo) [reply carsons/carsons.py]: yeah good point. Assigning the evaluations to variables makes it easier to read.

File: carsons/carsons.py (hunk 3 above)
Comment (veronicaguo) [reply carsons/carsons.py]: seems like it's exactly the same as in `CarsonsEquations` class... thinking of dropping it here

File: carsons/carsons.py [hunk 6]
```diff
@@ -183,3 +181,83 @@ def calculate_distance(positionᵢ, positionⱼ):
...
+            phase: self.GMR_cn(phase)
+            for phase in model.diameter_over_neutral.keys()
+        })
+        self.r.update({
+            phase: resistance / model.neutral_strand_count[phase]
+            for phase, resistance in model.neutral_strand_resistance.items()
+        })
+        return
+
+    def compute_d(self, i, j):
```
Comment (etimberg) [inline carsons/carsons.py]: # Comparison to Gridlab-d

To try and find the issues with the impedance calculations, I audited the gridlab-d source. I think this method differs from the [Gridlab-d](https://github.com/gridlab-d/gridlab-d/blob/master/powerflow/underground_line.cpp#L403-L450) distance calculations in the following cases.

## Background

Gridlab-d appears to use the following indexes for different phases. The snippets below use these phase indexes

| Index | Description |
| ----- | ----------- |
| 1 | Phase A |
| 2 | Phase B |
| 3 | Phase C |
| 4 | Concentric Neutral A |
| 5 | Concentric Neutral B |
| 6 | Concentric Neutral C |
| 7 | Neutral Conductor | 

## Conductor to Neutral Cable

### Python Implementation
In this case, the inputs are `I = {'A'}, J = {'N'}`

This leads to the following:

```python
I ^ J = {'A', 'N'}
I & J = {}

one_neutral_same_phase == False
different_phase == True
one_neutral == True
```
Thus the code passed the check on line 233 and returns `(distance_ij**2 + r**2) ** 0.5`.

### Gridlab-D Implementation
However, gridlab-d does the following in the case of A -> N:

```cpp
#define DIST(ph1, ph2) (has_phase(PHASE_##ph1) && has_phase(PHASE_##ph2) && config->line_spacing ? OBJECTDATA(config->line_spacing, line_spacing)->distance_##ph1##to##ph2 : 0.0)
D(1, 7) = DIST(A, N);
```

### Solution
I think the correct solution in this case is to return `distance_ij`

## Conductor to Own Concentric Neutral

## Python Implementation
In this case, the inputs are `I = {'A'}, J = {'A', 'N'}`

This leads to the following:

```python
I ^ J = {'N'}
I & J = {'A}

one_neutral_same_phase == True
different_phase == False
one_neutral == True
```

Thus the check on line 227 is true, and so `r` is returned. This value was calculated earlier on lines 200 and 201 as

```python
(diameter_over_neutral - model.neutral_strand_diameter[phase]) / 2
```

### Gridlab-d Distance
Gridlab-d does the following:
```cpp
dia_od1 = UG_GET(A, outer_diameter);
DIA(4) = UG_GET(A, neutral_diameter);
rad_14 = (dia_od1 - DIA(4)) / 24.0;
D(1, 4) = rad_14;
```
### Solution

Divide by 24 on line 201 instead of 2. Additionally, check that the gridlab-d outer diameter is the diameter over the neutral.

## Conductor to Different Phase Concentric Neutral

### Python Implementation

In this case, the inputs are `I = {'A'}, J = {'B', 'N'}`

This leads to the following:

```python
I ^ J = {'A', 'B', 'N'}
I & J = {}

one_neutral_same_phase == False
different_phase == True
one_neutral == True
```

Thus the check on line 233 is True, and so the python code returns `(distance_ij**2 + r**2) ** 0.5`

### Gridlab-d Implementation

```cpp
D(1, 5) = D(1, 2);
```

In otherwords, the distance from A -> BN is the same as the distance from A -> B.

## Concentric Neutral to Another Concentric Neutral

### Python Implementation

In this case, the inputs are `I = {'A', 'N'}, J = {'B', 'N'}`

This leads to the following:

```python
I ^ J = {'A', 'B'}
I & J = {'N'}

one_neutral_same_phase == False
different_phase == False
one_neutral == False
```

Thus the check on line 227 and the check on line 233 are False and so `distance_ij` is returned

### Gridlab-d Implementation
```cpp
D(4, 5) = D(1, 2);
```
In other words, the distance from AN -> BN is the same as the distance from A -> B. The code appears to be correct here

File: carsons/carsons.py [hunk 7]
```diff
@@ -183,3 +181,83 @@ def calculate_distance(positionᵢ, positionⱼ):
...
+        ΔX = Q_first_term * 2 + log(2)
+
+        if i == j:
+            X_o = -log(self.gmr[i]) - log(kᵢⱼ_Dᵢⱼ_ratio)
+        else:
+            X_o = -log(self.compute_d(i, j)) - log(kᵢⱼ_Dᵢⱼ_ratio)
+
+        return (X_o + ΔX) * self.ω * self.μ / (2 * π)
+
+    def GMR_cn(self, phase):
```
Comment (etimberg) [inline carsons/carsons.py]: I audited this against gridlab-d. The gridlab-d implementation is shown below

```cpp
GMRCN(4) = !(has_phase(PHASE_A) && strands_4 > 0) ? 0.0 : pow(GMR(4) * strands_4 * pow(rad_14, (strands_4 - 1)), (1.0 / strands_4));
```

## Differences
* Does not bail out early in the 0 case. Perhaps check `GMR_s`?
* `R` is incorrect per the comment on the distance function.

File: tests/test_concentric_neutral_cable.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    assert_array_almost_equal(
+        model.impedance,
+        array([
+            [1.2936 + 1j*0.6713, 0.4871 + 1j*0.2111, 0.4585 + 1j*0.1521],
+            [0.4871 + 1j*0.2111, 1.3022 + 1j*0.6326, 0.4871 + 1j*0.2111],
+            [0.4585 + 1j*0.1521, 0.4871 + 1j*0.2111, 1.2936 + 1j*0.6713],
+
+        ]) * OHM_PER_MILE_TO_OHM_PER_METER,
+        decimal=4
+    )
```
[the remark this answers is not in the document]
Comment (veronicaguo) [reply tests/test_concentric_neutral_cable.py]: Tested against OpenDSS for single/two phase cables

File: carsons/carsons.py (hunk 6 above)
Comment (veronicaguo) [reply carsons/carsons.py]: Thanks for the thorough audit @etimberg ! A couple points here:

- For case `Conductor to Own Concentric Neutral`, I think gridlab-d divides by `24` to include unit conversion from `inch` to `ft`. We handle everything using metric units, so dividing by `2` to get radius from diameter.

- For case `Conductor to Different Phase Concentric Neutral`, our implementation follows the assumption and example given in the Kersting's book, approximating the concentric neutrals as one single wire directly above their respective phase conductors. So there comes the Pythagorean theorem.

Comment (etimberg) [reply carsons/carsons.py]: That makes sense @veronicaguo re 2 instead of 24. For the `Conductor to Different Phase Concentric Neutral` case, I'm not sure that the distance calculation is correct. It assumes all the cables are in one horizontal plane because that's the only way the triangle becomes right angle. If the cables were arranged in a geometry similar to [this image](https://cdn.comsol.com/wordpress/2017/02/cross-section-of-underground-electrical-cable.png), the distance from the top conductor to one of the bottom concentric neutrals is formed by a non right triangle and so the extra term from the cosine law would need to be added.

Maybe a way to simplify the code here is to calculate the position (x,y) of the concentric neutral in a prior step. Then, all you'd do here is lookup `distance_ij`. It might be worthwhile testing this function independently too for the following cases:
* Horizontal geometry
* Vertical geometry
* Bundled like the image above

File: tests/test_concentric_neutral_cable.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+miles = ureg.miles
+ohms = ureg.ohms
+kft = ureg.feet * 1000
+
+
+def test_concentric_neutral_cable():
+    """
+    Validation test against example in Kersting's book.
+    """
+    model = ConcentricNeutralCarsonsEquations(ConcentricLineModel({
```
Comment (AnjoMan) [inline tests/test_concentric_neutral_cable.py]: I'm kind of souring on this API. I think its not great that you have to make two objects and then call a class method.
It might make more sense to either

# expose `impedence` as a function, and you do something like
``` python
 import carsons

z_abc = carsons.impedance(ConcentricNeutralCarsonsEquations(ConcentricLineModel(...)))
```
or we expose a function for each type of cable, e.g.

```python
import carsons
z_abc = carsons.concentric_cabe_impedance(ConcentricCableModel(...))
z_abc = carsons.line_impedance(LineModel(...))
```

Comment (veronicaguo) [reply tests/test_concentric_neutral_cable.py]: yeah good point! I like the first idea of exposing `imdedance` as a stand-alone function, so the users don't need to know about class method from `ConcentricNeutralCarsonsEquations`.

Comment (veronicaguo) [reply tests/test_concentric_neutral_cable.py]: done in 723ff87

File: carsons/carsons.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (veronicaguo) [reply carsons/carsons.py]: done in c1c5a13

File: carsons/carsons.py (hunk 3 above)
Comment (veronicaguo) [reply carsons/carsons.py]: dropped it here 53fa95f

File: carsons/carsons.py (hunk 4 above)
Comment (veronicaguo) [reply carsons/carsons.py]: exposed `impedance` as a stand-alone function in 723ff87
